### PR TITLE
tcp: add uv_tcp_close_reset method

### PR DESCRIPTION
**Pull request by santigimeno**
It resets a TCP connection by sending a RST packet. Due to some platform
inconsistencies, mixing of `uv_shutdown` and `uv_tcp_close_reset` calls
is not allowed.

Fixes: https://github.com/libuv/libuv/issues/1991

### CODE REVIEW

File: docs/src/tcp.rst [hunk 1]
```diff
@@ -113,3 +113,11 @@ API
         mapping
 
 .. seealso:: The :c:type:`uv_stream_t` API functions also apply.
+
+.. c:function:: int uv_tcp_close_reset(uv_tcp_t* handle, uv_close_cb close_cb)
+
+    It resets a TCP connection by sending a RST packet. Due to some platform
+    inconsistencies, mixing of :c:func:`uv_shutdown` and
+    :c:func:`uv_tcp_close_reset` calls is not allowed.
```
Comment (addaleax) [inline docs/src/tcp.rst]: I’d make it explicit that this calls `uv_close()`.

File: src/win/stream.c [hunk 2]
```diff
@@ -198,15 +198,18 @@ int uv_try_write(uv_stream_t* stream,
 int uv_shutdown(uv_shutdown_t* req, uv_stream_t* handle, uv_shutdown_cb cb) {
   uv_loop_t* loop = handle->loop;
 
-  if (!(handle->flags & UV_HANDLE_WRITABLE)) {
-    return UV_EPIPE;
+  if (!(handle->flags & UV_HANDLE_WRITABLE) ||
+      handle->flags & UV_HANDLE_SHUTTING ||
+      uv__is_closing(handle)) {
+    return UV_ENOTCONN;
```
Comment (santigimeno) [inline src/win/stream.c]: Not completely sure about this change though it makes it consistent across platforms.

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1531/

File: docs/src/tcp.rst (hunk 1 above)
Comment (santigimeno) [reply docs/src/tcp.rst]: Docs updated. Thanks!

File: src/unix/tcp.c [hunk 3]
```diff
@@ -308,6 +308,21 @@ int uv_tcp_getpeername(const uv_tcp_t* handle,
 }
 
 
+int uv_tcp_close_reset(uv_tcp_t* handle, uv_close_cb close_cb) {
+  struct linger l = { 1, 0 };
```
Comment (bnoordhuis) [inline src/unix/tcp.c]: This contradicts the docs ("a linger interval of zero.")

File: test/test-tcp-close-reset.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+
+static void do_write(uv_tcp_t* handle) {
+  uv_write_t* req;
+  uv_buf_t buf;
+  int i, r;
+
+  buf = uv_buf_init("PING", 4);
+  for (i = 0; i < NUM_WRITE_REQS; i++) {
+    req = malloc(sizeof *req);
```
Comment (bnoordhuis) [inline test/test-tcp-close-reset.c]: I'd use a `static uv_write_t write_reqs[4];` for this and then check `i < ARRAY_SIZE(write_reqs)`, or do:
```c
for (req = write_reqs; req < ARRAY_END(write_reqs); req++) {
  /* ... */
}
```

If we don't have an `ARRAY_END()` macro already, its definition is this:
```c
#define ARRAY_END(x) ((x) + ARRAY_SIZE(x))
```

File: test/test-tcp-close-reset.c [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    ASSERT(0 == uv_shutdown(&shutdown_req, (uv_stream_t*) handle, shutdown_cb));
+    ASSERT(UV_EINVAL == uv_tcp_close_reset(handle, close_cb));
+  } else {
+    ASSERT(0 == uv_tcp_close_reset(handle, close_cb));
+    ASSERT(UV_ENOTCONN == uv_shutdown(&shutdown_req, (uv_stream_t*) handle, shutdown_cb));
+  }
+}
+
+static void alloc_cb(uv_handle_t* handle, size_t size, uv_buf_t* buf) {
+  buf->base = malloc(size);
```
Comment (bnoordhuis) [inline test/test-tcp-close-reset.c]: You could use a `static char slab[1024];` here.

File: test/test-tcp-close-reset.c [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+  else
+    ASSERT(req->handle == (uv_stream_t*) &tcp_accepted);
+
+  shutdown_cb_called++;
+}
+
+
+static void read_cb(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT((uv_tcp_t*)stream == &tcp_accepted);
+  read_size += nread;
```
Comment (bnoordhuis) [inline test/test-tcp-close-reset.c]: This looks wrong. `nread` can be < 0. You handle `UV_EOF` buf if it's any other error, the test is going to act wonky (time out, I suspect.)

**Comment by saghul** [comment]
Why does this function also close the handle? I know there is not much one can do afterwards, but if we had `uv_tcp_reset(handle)` then the user would know that, as usual they need to call `uv_close`.

File: docs/src/tcp.rst [hunk 7]
```diff
@@ -113,3 +113,13 @@ API
         mapping
 
 .. seealso:: The :c:type:`uv_stream_t` API functions also apply.
+
+.. c:function:: int uv_tcp_close_reset(uv_tcp_t* handle, uv_close_cb close_cb)
+
+    It resets a TCP connection by sending a RST packet. This is accomplished by
```
Comment (cjihrig) [inline docs/src/tcp.rst]:
```suggestion
    Resets a TCP connection by sending a RST packet. This is accomplished by
```

File: docs/src/tcp.rst [hunk 8]
```diff
@@ -113,3 +113,13 @@ API
...
+    setting the `SO_LINGER` socket option with a linger interval of zero and
+    then calling to :c:func:`uv_close`.
```
Comment (cjihrig) [inline docs/src/tcp.rst]:
```suggestion
    then calling :c:func:`uv_close`.
```

File: src/unix/tcp.c [hunk 9]
```diff
@@ -308,6 +308,21 @@ int uv_tcp_getpeername(const uv_tcp_t* handle,
...
+
+  /* Disallow setting SO_LINGER to zero due to some platform inconsistencies */
+  if (handle->flags & UV_HANDLE_SHUTTING)
+    return UV_EINVAL;
+
+  if (setsockopt(uv__stream_fd(handle), SOL_SOCKET, SO_LINGER, &l, sizeof(l)))
```
Comment (cjihrig) [inline src/unix/tcp.c]: `if (0 != ...)`

File: src/win/tcp.c [hunk 10]
```diff
@@ -549,6 +549,21 @@ static void uv_tcp_queue_read(uv_loop_t* loop, uv_tcp_t* handle) {
...
 
 
+int uv_tcp_close_reset(uv_tcp_t* handle, uv_close_cb close_cb) {
+  struct linger l = { 1, 0 };
+
+  /* Disallow setting SO_LINGER to zero due to some platform inconsistencies */
+  if (handle->flags & UV_HANDLE_SHUTTING)
+    return UV_EINVAL;
+
+  if (setsockopt(handle->socket, SOL_SOCKET, SO_LINGER, &l, sizeof(l)))
```
Comment (cjihrig) [inline src/win/tcp.c]: Same comment here about checking the return value explicitly.

**Comment by santigimeno** [comment]
> Why does this function also close the handle? I know there is not much one can do afterwards, but if we had uv_tcp_reset(handle) then the user would know that, as usual they need to call uv_close.

For the RST packet to be sent, you need to call to `close()` so I think it makes sense calling `uv_close()`, otherwise the function would just set the socket option but would not reset the connection. Does this answer your question or am I missing something?

File: test/test-tcp-close-reset.c (hunk 4 above)
Comment (santigimeno) [reply test/test-tcp-close-reset.c]: Yes, I was lazy and copied part of the code from other tests. Will improve it per your suggestions.

**Comment by santigimeno** [comment]
Updated with all the suggested changes. Thanks!
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1536/

**Comment by vtjnash** [comment]
> For the RST packet to be sent, you need to call to close()

I guess the alternative is to be more like `uv_shutdown` and call `close` but don't close the libuv handle yet? But there, the API lets you half-close it, so it makes more sense that you may continue using (the other half of) it afterwards. Whereas after this, there's not much left that can be done useful, so `uv_close` makes sense to me.

**Comment by saghul** [comment]
Ah, got it! 👍 then!

File: test/test-tcp-close-reset.c [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <errno.h>
+#include <string.h> /* memset */
+
+static uv_loop_t* loop;
+static uv_tcp_t tcp_server;
```
Comment (bnoordhuis) [inline test/test-tcp-close-reset.c]: This handle is initialized but never closed, AFAICT.

Comment (santigimeno) [reply test/test-tcp-close-reset.c]: You're right. It should be fixed now. Thanks!

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1539/

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/958e85fbd842be06da8484305996f42149ba5039. Thanks!

**Comment by vtjnash** [comment]
Just curious if anyone tested this on AIX? Attempting to use this in #2409 for its stated purpose (using to test the error path of libuv code) seems to be failing CI there, but seems to be working on all other platforms.

**Pull request closed without merging**